pkg/services/screenshot/singleflight.go

```diff
@@ -0,0 +1,47 @@
+package screenshot
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/singleflight"
+)
+
+type captureFunc func(ctx context.Context, opts ScreenshotOptions) (*Screenshot, error)
+
+// SingleFlight is an interface for single flighters. It prevents multiple screenshots of the same scene
+// from being taken at the same time, instead taking the screenshot once and sharing it with all callers
+// that requested the same scene.
+//
+//go:generate mockgen -destination=singleflight_mock.go -package=screenshot github.com/grafana/grafana/pkg/services/screenshot SingleFlight
+type SingleFlight interface {
+	// Do returns a screenshot for the options or an error. It prevents multiple screenshots with
+	// the same options from being taken at the same time. Instead, the screenshot is taken once
+	// and shared with all callers that requested a screenshot with the same options.
+	Do(ctx context.Context, opts ScreenshotOptions, fn captureFunc) (*Screenshot, error)
+}
+
+type SingleFlighter struct {
+	f singleflight.Group
+}
+
+func NewSingleFlight() SingleFlight {
+	return &SingleFlighter{}
+}
+
+// Screenshot returns a screenshot or an error. It ensures that at most one screenshot
+// can be taken at a time for the same dashboard and theme. Duplicate screenshots
+// wait for the first screenshot to complete and receive the same screenshot.
+func (s *SingleFlighter) Do(ctx context.Context, opts ScreenshotOptions, fn captureFunc) (*Screenshot, error) {
+	k := fmt.Sprintf("%s-%d-%s", opts.DashboardUID, opts.PanelID, opts.Theme)
+
+	v, err, _ := s.f.Do(k, func() (interface{}, error) {
```

Comment:
Why don't we care about this value? If the result is shared, we might do duplicate work in each caller, right?

Reply:
You're right, I think we should! For example, we would want to know about it to avoid uploading the same image more than once, and even saving the metadata to the database multiple times.

Reply:
I made some changes following this feedback!

1. I moved SingleFlight from screenshot package to the image package so we can do a singeflight of not just taking the screenshot, but also uploading it and saving it to the database
2. I removed the use of (but not the code) the cache because I don't think it makes sense to cache the image across evaluations, just across alert instances in the same evaluation